cpp/src/gandiva/decimal_xlarge.cc

```diff
@@ -0,0 +1,156 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Operations that can deal with very large values (256-bit).
+//
+// The intermediate results with decimal can be larger than what can fit into 128-bit,
+// but the final results can fit in 128-bit after scaling down. These functions deal
+// with operations on the intermediate values.
+//
+
+#include "gandiva/decimal_xlarge.h"
+
+#include <boost/multiprecision/cpp_int.hpp>
+#include <limits>
+#include <vector>
+
+#include "arrow/util/basic_decimal.h"
+#include "gandiva/decimal_type_util.h"
+#include "gandiva/logging.h"
+
+#ifndef GANDIVA_UNIT_TEST
+#include "gandiva/engine.h"
+#include "gandiva/exported_funcs.h"
+
+namespace gandiva {
+
+void ExportedDecimalFunctions::AddMappings(Engine* engine) const {
+  std::vector<llvm::Type*> args;
+  auto types = engine->types();
+
+  // gdv_multiply_and_scale_down
+  args = {types->i64_type(),      // int64_t x_high
+          types->i64_type(),      // uint64_t x_low
+          types->i64_type(),      // int64_t y_high
+          types->i64_type(),      // uint64_t x_low
+          types->i32_type(),      // int32_t reduce_scale_by
+          types->i64_ptr_type(),  // int64_t* out_high
+          types->i64_ptr_type(),  // uint64_t* out_low
+          types->i8_ptr_type()};  // bool* overflow
+
+  engine->AddGlobalMappingForFunc("gdv_multiply_and_scale_down",
```

Comment:
What is the purpose of this declaration? AFAICT, `gdv_multiply_and_scale_down` is only called from `MultiplyMaxPrecision` which is a regular C++ function.

I find the current call path quite convoluted, and it's not obvious what will be compiled to LLVM bitcode and what will be compiled to native code ahead of time.

Reply:
only files in the folder src/gandiva/precompiled and basic_decimal.cc are compiled to llvm bitcode ahead of time. The rest are compiled to native code.

In the files that are precompiled to llvm bitcode, I'm avoiding references to libstdc++/boost since it complicates symbol resolution at run-time. In such cases, I add a C wrapper in non-precompiled-to-llvm code (i.e outside src/gandiva/precompiled), and export it as a global function in JIT.

I'll change the name of this to gdv_xlarge_multiply_and_scale_down and add a comment here on why it can't be precompiled-to-bitcode.